core/src/main/scala/org/apache/spark/util/SizeEstimator.scala

```diff
@@ -316,62 +338,40 @@ object SizeEstimator extends Logging {
    */
   private def getClassInfo(cls: Class[_]): ClassInfo = {
     // Check whether we've already cached a ClassInfo for this class
-    val info = classInfos.get(cls)
+    val info = classInfos.get().get(cls)
     if (info != null) {
       return info
     }
 
     val parent = getClassInfo(cls.getSuperclass)
+    val fields = cls.getDeclaredFields
+    val fieldCount = fields.length
     var shellSize = parent.shellSize
-    var pointerFields = parent.pointerFields
-    val sizeCount = Array.fill(fieldSizes.max + 1)(0)
+    var fieldOffsets = parent.fieldOffsets.toList
+
+    var index = 0
 
-    // iterate through the fields of this class and gather information.
-    for (field <- cls.getDeclaredFields) {
+    while (index < fieldCount) {
+      val field = fields(index)
       if (!Modifier.isStatic(field.getModifiers)) {
         val fieldClass = field.getType
         if (fieldClass.isPrimitive) {
-          sizeCount(primitiveSize(fieldClass)) += 1
+          if (cls == classOf[Double] || cls == classOf[Long]) {
```

Comment:
This and the logic changes below aren't obviously OK. this seems to lose a lot of logic. I think this has to be explained or backed out